### PR TITLE
nixos/waybar: use upstream service file

### PR DESCRIPTION
**Pull request by m-bdf**
###### Description of changes

Use the more complete upstream [waybar.service](https://github.com/Alexays/Waybar/blob/master/resources/waybar.service.in) file instead of manually re-implementing it.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rodrgz** [comment]
Result of `nixpkgs-review pr 242728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>waybar</li>
  </ul>
</details>

File: nixos/modules/programs/wayland/waybar.nix [hunk 1]
```diff
@@ -13,11 +13,9 @@ in
...
   config = mkIf cfg.enable {
     environment.systemPackages = [ cfg.package ];
-    systemd.user.services.waybar = {
-      description = "Waybar as systemd service";
-      wantedBy = [ "graphical-session.target" ];
-      partOf = [ "graphical-session.target" ];
-      script = "${cfg.package}/bin/waybar";
+    systemd = {
+      packages = [ cfg.package ];
+      user.services.waybar.wantedBy = [ "graphical-session.target" ];
```
Comment (SuperSandro2000) [inline nixos/modules/programs/wayland/waybar.nix:18]: Isn't that already here https://github.com/Alexays/Waybar/blob/master/resources/waybar.service.in#L13-L14 ?

File: pkgs/applications/misc/waybar/default.nix [hunk 2]
```diff
@@ -65,28 +68,17 @@ stdenv.mkDerivation rec {
     )
   '';
 
-  nativeBuildInputs = [
-    meson ninja pkg-config scdoc wrapGAppsHook
-  ] ++ lib.optional withMediaPlayer gobject-introspection;
+  nativeBuildInputs = [ meson ninja pkg-config scdoc wrapGAppsHook systemd ]
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/waybar/default.nix:71]: Why systemd in nativeBuildInputs?

File: pkgs/applications/misc/waybar/default.nix [hunk 3]
```diff
@@ -65,28 +68,17 @@ stdenv.mkDerivation rec {
...
+    ++ lib.optional withMediaPlayer gobject-introspection;
 
-  propagatedBuildInputs = lib.optionals withMediaPlayer [
-    glib
-    playerctl
-    python3.pkgs.pygobject3
-  ];
+  propagatedBuildInputs = lib.optionals withMediaPlayer [ glib playerctl python3.pkgs.pygobject3 ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/waybar/default.nix:74]: We shouldn't change the formatting here, one item per line is actually better git diffable.

File: nixos/modules/programs/wayland/waybar.nix (hunk 1 above)
Comment (m-bdf) [reply nixos/modules/programs/wayland/waybar.nix:18]: Yes but the Install section is not respected with `systemd.packages` (#81138)

File: pkgs/applications/misc/waybar/default.nix (hunk 2 above)
Comment (m-bdf) [reply pkgs/applications/misc/waybar/default.nix:71]: It's used in [meson.build](https://github.com/Alexays/Waybar/blob/effad1a5c3111fa8fecf44b6129772c7cd144990/meson.build#L147) to retrieve the user units directory to put the waybar.service file into.

File: pkgs/applications/misc/waybar/default.nix (hunk 3 above)
Comment (m-bdf) [reply pkgs/applications/misc/waybar/default.nix:74]: That was just for consistency, should I then change all the other lists in this file?

Comment (SuperSandro2000) [reply pkgs/applications/misc/waybar/default.nix:74]: if that's important to you, yes

File: pkgs/applications/misc/waybar/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/misc/waybar/default.nix:71]: maybe we add a short comment here as this is a bid unusual

Comment (m-bdf) [reply pkgs/applications/misc/waybar/default.nix:71]: It's only used at compile time, which from my understanding is what distinguishes nativeBuildInputs and buildInputs (also used at run time), but I may be wrong?

Comment (SuperSandro2000) [reply pkgs/applications/misc/waybar/default.nix:71]: Yes, nativeBuildInputs is for programs running on the build machine and buildInputs is for the target machine. Runtime dependencies are detected by nix.

Comment (m-bdf) [reply pkgs/applications/misc/waybar/default.nix:71]: I'm not sure where to put it then, it's never actually run on any machine, only used by meson to get the pkg-config variable that I set [a few lines below](https://github.com/NixOS/nixpkgs/blob/3d27246376b85d39bee6dda982e9ee6d89cf481b/pkgs/applications/misc/waybar/default.nix#L123)…

**Review by minijackson — APPROVED** [review]
Looks good, thanks!

---

Result of `nixpkgs-review pr 242728` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>waybar</li>
  </ul>
</details>

File: pkgs/by-name/wa/waybar/package.nix [hunk 4]
```diff
@@ -153,6 +155,7 @@ stdenv.mkDerivation (finalAttrs: {
       "cava" = cavaSupport;
       "dbusmenu-gtk" = traySupport;
       "jack" = jackSupport;
+      "libevdev" = evdevSupport;
```
Comment (khaneliman) [inline pkgs/by-name/wa/waybar/package.nix:158]: this is already in the derivation... conflict resolution issue / rebase needed? https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/wa/waybar/package.nix#L156

File: pkgs/by-name/wa/waybar/package.nix [hunk 5]
```diff
@@ -163,13 +166,15 @@ stdenv.mkDerivation (finalAttrs: {
       "pulseaudio" = pulseSupport;
       "rfkill" = rfkillSupport;
       "sndio" = sndioSupport;
-      "systemd" = false;
+      "systemd" = true;
```
Comment (khaneliman) [inline pkgs/by-name/wa/waybar/package.nix:169]: should probably follow the same theme as the other flags if someone wants to build without systemd dependency

**Review by khaneliman — CHANGES_REQUESTED** [review]

File: pkgs/by-name/wa/waybar/package.nix (hunk 5 above)
Comment (m-bdf) [reply pkgs/by-name/wa/waybar/package.nix:169]: This option only enabled copying `waybar.service` to the correct location (`PKG_CONFIG_SYSTEMD_SYSTEMDUSERUNITDIR`) during the build, nothing else.

The result waybar binary remains identical as the systemd dependency is never used at runtime, so I don't really see a need to allow disabling it. I even believe a `systemdSupport` option would actually confuse people by making them think the waybar binary can depend on and use the systemd dependency at runtime.

Comment (eclairevoyant) [reply pkgs/by-name/wa/waybar/package.nix:171]: That's what `buildInputs` (i.e. `depsHostTarget`) implies, that it's a runtime linkage.
If it's not used at runtime why would it be there?

**Pull request closed without merging**